app/controllers/UnloadingGuidanceController.scala

```diff
@@ -17,29 +17,53 @@
 package controllers
 
 import controllers.actions._
+import models.P5.ArrivalMessageType.UnloadingPermission
 import models.{ArrivalId, NormalMode}
+import pages.{GoodsTooLargeForContainerYesNoPage, NewAuthYesNoPage}
 import play.api.i18n.{I18nSupport, MessagesApi}
 import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.P5.UnloadingPermissionMessageService
 import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import viewModels.UnloadingGuidanceViewModel.UnloadingGuidanceViewModelProvider
 import views.html.UnloadingGuidanceView
 
 import javax.inject.Inject
+import scala.concurrent.ExecutionContext
 
 class UnloadingGuidanceController @Inject() (
   override val messagesApi: MessagesApi,
   val controllerComponents: MessagesControllerComponents,
   actions: Actions,
-  view: UnloadingGuidanceView
-) extends FrontendBaseController
+  view: UnloadingGuidanceView,
+  unloadingPermission: UnloadingPermissionMessageService,
+  getMandatoryPage: SpecificDataRequiredActionProvider,
+  unloadingGuidanceViewModel: UnloadingGuidanceViewModelProvider
+)(implicit executionContext: ExecutionContext)
+    extends FrontendBaseController
     with I18nSupport {
 
-  def onPageLoad(arrivalId: ArrivalId, messageId: String): Action[AnyContent] =
-    actions.requireData(arrivalId) {
-      implicit request =>
-        Ok(view(request.userAnswers.mrn, arrivalId, messageId))
-    }
+  def onPageLoad(arrivalId: ArrivalId): Action[AnyContent] =
+    actions
+      .requireData(arrivalId)
+      .andThen(getMandatoryPage(NewAuthYesNoPage))
+      .async {
+        implicit request =>
+          val newAuth: Boolean               = request.arg
+          val goodsTooLarge: Option[Boolean] = request.userAnswers.get(GoodsTooLargeForContainerYesNoPage)
+
+          val message = unloadingPermission.getMessageId(arrivalId, UnloadingPermission)
+
+          for {
+            messageId <- message
+          } yield messageId match {
+            case Some(id) =>
+              Ok(view(request.userAnswers.mrn, arrivalId, id, NormalMode, unloadingGuidanceViewModel.apply(newAuth, goodsTooLarge)))
+            case None => Redirect(controllers.routes.ErrorController.technicalDifficulties())
+          }
```

Comment:
Think this is ok rather than saving the message ID in user answers. If we suddenly need to start using this everywhere, it might be worth revisiting that, or at least sticking it in action. For now though this is purely stylistic but when we're only mapping over one future the for-yield isn't much use. Instead we could just have

```
unloadingPermission.getMessageId(arrivalId, UnloadingPermission).map {
  case Some(messageId) =>
    Ok(view(request.userAnswers.mrn, arrivalId, id, NormalMode, unloadingGuidanceViewModel.apply(newAuth, goodsTooLarge)))
    case None => Redirect(controllers.routes.ErrorController.technicalDifficulties())
}
```

Reply:
Done